status/controller.go

```diff
@@ -6,49 +6,166 @@ import (
+	o := c.forObject.DeepCopyObject().(Object)
+	gvk := object.GVK(o)
+	objectLabels := prometheus.Labels{
+		MetricLabelGroup:     gvk.Group,
+		MetricLabelKind:      gvk.Kind,
+		MetricLabelNamespace: req.Namespace,
+		MetricLabelName:      req.Name,
 	}
-	ConditionCount.DeletePartialMatch(labels)
 
 	if err := c.kubeClient.Get(ctx, req.NamespacedName, o); err != nil {
-		return reconcile.Result{}, client.IgnoreNotFound(err)
+		if errors.IsNotFound(err) {
+			ConditionCount.DeletePartialMatch(objectLabels)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("getting object, %w", err)
 	}
 
+	currentConditions := o.StatusConditions()
+	observedConditions := c.observedConditions[req]
+	c.observedConditions[req] = currentConditions
+
+	// Detect and record condition counts
 	for _, condition := range o.GetConditions() {
-		ConditionCount.MustCurryWith(labels).With(prometheus.Labels{
-			metricLabelConditionType:   condition.Type,
-			metricLabelConditionStatus: string(condition.Status),
+		ConditionCount.MustCurryWith(objectLabels).With(prometheus.Labels{
+			MetricLabelConditionType:   string(condition.Type),
+			MetricLabelConditionStatus: string(condition.Status),
 		}).Set(1)
 	}
+	for _, observedCondition := range observedConditions.List() {
+		if currentCondition := currentConditions.Get(observedCondition.Type); currentCondition == nil || currentCondition.Status != observedCondition.Status {
+			ConditionCount.MustCurryWith(objectLabels).Delete(prometheus.Labels{
+				MetricLabelConditionType:   string(observedCondition.Type),
+				MetricLabelConditionStatus: string(observedCondition.Status),
+			})
+		}
+	}
+
+	// Detect and record status transitions. This approach is best effort,
+	// since we may batch multiple writes within a single reconcile loop.
+	// It's exceedingly difficult to atomically track all changes to an
+	// object, since the Kubernetes is evenutally consistent by design.
+	// Despite this, we can catch the majority of transition by remembering
+	// what we saw last, and reporting observed changes.
+	//
+	// We rejected the alternative of tracking these changes within the
+	// condition library itself, since you cannot guarantee that a
+	// transition made in memory was successfully persisted.
+	//
+	// Automatic monitoring systems must assume that these observations are
+	// lossy, specifically for when a condition transition rapidly. However,
+	// for the common case, we want to alert when a transition took a long
+	// time, and our likelyhood of observing this is much higher.
+	for _, condition := range currentConditions.List() {
+		observedCondition := observedConditions.Get(condition.Type)
+		if observedCondition == nil || observedCondition.GetStatus() == condition.GetStatus() {
+			continue
+		}
+		duration := condition.LastTransitionTime.Time.Sub(observedCondition.LastTransitionTime.Time).Seconds()
+		ConditionDuration.MustCurryWith(objectLabels).With(prometheus.Labels{
```

Comment:
How big do our condition duration metrics get over the lifetime of the controller? We obviously shouldn't delete them after the object is gone, since we still need to scrape them to get accurate metrics over a given point of time; however, does this cardinality ever become a problem when it comes to memory usage since we are never deleting these counters?

Reply:
I suspect that many metrics have this problem. This problem exists whether or not we have the cardinality, since we are tracking individual data points for every transition. Every transition that occurs will grow the memory. I wonder if prometheus has a concept of whether or not a metric has been scraped such that the source (our process) can flush the memory. Maybe there's a TTL or something that we could use .

Reply:
Discussed offline. This is tricky, and it's not totally clear how histograms are aggregated over time. e.g., if we clear out a histogram metric, will that cause a spike in data? If the process restarts, is that equivalent to clearing it out? We decided to address this in a followon and remove NN from the duration metric for now.